tests/mocks/services.py

```diff
@@ -270,3 +312,147 @@ async def get_position(self, *, timeout: Optional[float] = None) -> Pose:
 
     async def do_command(self, command: Mapping[str, ValueTypes], *, timeout: Optional[float] = None, **kwargs) -> Mapping[str, ValueTypes]:
         return {"command": command}
+
+
+class MockData(DataServiceBase):
+    def __init__(
+        self,
+        tabular_response: List[Mapping[str, Any]],
+        binary_response: List[bytes],
+        delete_remove_response: int,
+        tags_response: List[str],
+        bbox_labels_response: List[str]
+    ):
+        self.tabular_response = tabular_response
+        self.binary_response = binary_response
+        self.delete_remove_response = delete_remove_response
+        self.tags_response = tags_response
+        self.bbox_labels_response = bbox_labels_response
+        self.was_tabular_data_requested = False
+        self.was_binary_data_requested = False
+
+    async def TabularDataByFilter(self, stream: Stream[TabularDataByFilterRequest, TabularDataByFilterResponse]) -> None:
+        request = await stream.recv_message()
+        assert request is not None
+        if self.was_tabular_data_requested:
+            await stream.send_message(TabularDataByFilterResponse(data=None))
+            return
+        self.filter = request.data_request.filter
+        n = len(self.tabular_response)
+        tabular_response_structs = [None] * n
+        for i in range(n):
+            s = Struct()
+            s.update(self.tabular_response[i])
+            tabular_response_structs[i] = s
+        await stream.send_message(TabularDataByFilterResponse(data=[TabularData(data=struct) for struct in tabular_response_structs]))
+        self.was_tabular_data_requested = True
+
+    async def BinaryDataByFilter(self, stream: Stream[BinaryDataByFilterRequest, BinaryDataByFilterResponse]) -> None:
+        request = await stream.recv_message()
+        assert request is not None
+        if self.was_binary_data_requested:
+            await stream.send_message(BinaryDataByFilterResponse())
+            return
+        self.filter = request.data_request.filter
+        await stream.send_message(BinaryDataByFilterResponse(data=[BinaryData(binary=binary_data) for binary_data in self.binary_response]))
+        self.was_binary_data_requested = True
```

Comment:
Do you mean replacing line 358 (self.was_binary_data_requested = True) with an empty call to send_message? I haven't been able to send_message() more than once per request received (i.e., I can only do it once per call to MockData.Tabular/BinaryDataByFilter). I think this is because it's a unaryunary gRPC method, so the only thing I can think of to make sure this request doesn't hang is including this flag at a higher scope to "keep track" of incoming requests.

Reply:
Got it, thanks for clarifying! Will resolve.